File: workers/runeToEmbed.js
import { MessageEmbed, MessageButton, MessageActionRow } from 'discord.js';
import {chunkArrayinGroupsWithCB} from './arrayHelper.js';
import { genInfoLink } from './runes.js';

export function runeToMessage (inputRuneObject) {
  let output = {
    ephemeral: true,
    embeds : [runeToEmbed(inputRuneObject)],
    components : [componentRowWith(runeLinkButton(inputRuneObject.name, inputRuneObject.descURL))]
  }
  return output;
}

export function allRunesLinks(futharkArray) {
  // hard limit of five by five here
  // five Action Rows, with Five Buttons each
  let output = {
    ephemeral: true,
    components : []
  };
  let runeNameMatrix = chunkArrayinGroupsWithCB(futharkArray, 5);
  runeNameMatrix.forEach(nameArray => {
    let messageRow = new MessageActionRow();
    nameArray.forEach(runeName => {
      messageRow.addComponents(runeNameButton(runeName));
    });
    output.components.push(messageRow);
  });
  return output;
}

function runeToEmbed(runeObject) {
  const embed = new MessageEmbed({
    title : runeObject.name,
    url : runeObject.descURL,
    image : {
      "url": runeObject.imgURL
    },
    description : `You drew ${runeObject.name}`
  });

  return embed;
}

function componentRowWith(inputButton) {
  let outputRow = new MessageActionRow()
    .addComponents(inputButton);
  return outputRow;
}

function runeLinkButton (inputTitle, URLinput) {
  let outputButton = new MessageButton()
    .setLabel(`Go to ${inputTitle} info page.`)
    .setURL(URLinput)
    .setStyle('LINK');
  return outputButton;
}

function runeNameButton(name) {
  let outputButton = new MessageButton()
  .setLabel(name.toString())
  .setURL(genInfoLink(name))
  .setStyle('LINK');
return outputButton;
}

File: workers/runeToEmbed.go
package workers

import (
	"github.com/bwmarrin/discordgo"
)

// RuneToMessage converts a rune to a Discord message.
func RuneToMessage(runeObj Rune) *discordgo.MessageSend {
	return &discordgo.MessageSend{
		Embeds: []*discordgo.MessageEmbed{runeToEmbed(runeObj)},
		Components: []discordgo.MessageComponent{
			componentRowWith(runeLinkButton(runeObj.Name, runeObj.DescURL)),
		},
	}
}

func runeToEmbed(runeObj Rune) *discordgo.MessageEmbed {
	return &discordgo.MessageEmbed{
		Title:       runeObj.Name,
		URL:         runeObj.DescURL,
		Description: "You drew " + runeObj.Name,
		Image: &discordgo.MessageEmbedImage{
			URL: runeObj.ImgURL,
		},
	}
}

func componentRowWith(button discordgo.Button) discordgo.ActionsRow {
	return discordgo.ActionsRow{
		Components: []discordgo.MessageComponent{button},
	}
}

func runeLinkButton(title, url string) discordgo.Button {
	return discordgo.Button{
		Label: "Go to " + title + " info page.",
		Style: discordgo.LinkButton,
		URL:   url,
	}
}

func AllRunesLinks(futharkArray []string) *discordgo.MessageSend {
	output := &discordgo.MessageSend{
		Components: []discordgo.MessageComponent{},
	}
	runeNameMatrix := ChunkArray(futharkArray, 5)
	for _, nameArray := range runeNameMatrix {
		messageRow := discordgo.ActionsRow{}
		for _, runeName := range nameArray {
			messageRow.Components = append(messageRow.Components, runeNameButton(runeName))
		}
		output.Components = append(output.Components, messageRow)
	}
	return output
}

func runeNameButton(name string) discordgo.Button {
	return discordgo.Button{
		Label: name,
		Style: discordgo.LinkButton,
		URL:   genInfoLink(name),
	}
}